### PR TITLE
Add Icon to wxMimeTypesManager in GTK under UNIX

### PR DESCRIPTION
**Pull request by Hanmac**
this is the development of: http://trac.wxwidgets.org/ticket/17536

with improving the Unix wxMimeTypesManagerImpl to get an icon location from giving mimetype.
then adding a GTK implementing that (might be done for other ports later too)

updating the samples to show that icon location.
(might be interesting if the icon should be shown in a wxStaticBitmap too)

for the bakefiles it might be the best to do a GTK_UNIX_SRC variable because the others already have Icon stuff.

means the PR is not complete yet, but i would like it if you others would help me finish it.

### CODE REVIEW

File: include/wx/gtk/mimetype.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        wx/gtk/mimetype.h
+// Purpose:     classes and functions to manage MIME types
+// Author:      Hans Mackowiak
+// Modified by:
```
Comment (vadz) [inline include/wx/gtk/mimetype.h]: We shouldn't have these "Modified" lines in new code.

File: include/wx/gtk/mimetype.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Created:     05.06.2016
```
Comment (vadz) [inline include/wx/gtk/mimetype.h]: Could you please use ISO 8601 format for consistency, i.e. 2016-06-05?

File: include/wx/gtk/mimetype.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// Copyright:   (c) 2016 Hans Mackowiak <hanmac@gmx.de>
+// Licence:     wxWindows licence (part of wxExtra library)
```
Comment (vadz) [inline include/wx/gtk/mimetype.h]: Not "wxExtra".

File: include/wx/gtk/mimetype.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef _GTK_MIMETYPE_IMPL_H
```
Comment (vadz) [inline include/wx/gtk/mimetype.h]: Should have `_WX` prefix.

File: include/wx/gtk/mimetype.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+protected:
+    wxString GetIconFromMimeType(const wxString& mime) wxOVERRIDE;
+};
+
+
+class WXDLLIMPEXP_CORE wxGTKMimeTypesManagerFactory : public wxMimeTypesManagerFactory
+{
+public:
+    wxGTKMimeTypesManagerFactory() {}
+    virtual ~wxGTKMimeTypesManagerFactory() {}
```
Comment (vadz) [inline include/wx/gtk/mimetype.h]: Seems unnecessary?

File: src/gtk/app.cpp [hunk 6]
```diff
@@ -31,6 +31,9 @@
 #include <gtk/gtk.h>
 #include "wx/gtk/private.h"
 
+#if wxUSE_MIMETYPE
```
Comment (vadz) [inline src/gtk/app.cpp]: There is also check for this inside `wx/gtk/mimetype.h`, one of them seems unnecessary.

File: src/gtk/mimetype.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+#ifdef __BORLANDC__
+    #pragma hdrstop
+#endif
+
+#if wxUSE_MIMETYPE && wxUSE_FILE
+
+#include "wx/gtk/mimetype.h"
+
+#include <gio/gio.h>
```
Comment (vadz) [inline src/gtk/mimetype.cpp]: Can we be sure that it's always available? Or could we need `PKG_CHECK_MODULES(gio)` in configure?

File: src/gtk/mimetype.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+#include <gtk/gtk.h>
+
+#include "wx/gtk/private/string.h"
+
+wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString& mime)
+{
+    wxString icon;
+    wxGtkString type(g_content_type_from_mime_type(mime.mb_str()));
```
Comment (vadz) [inline src/gtk/mimetype.cpp]: Should be `.utf8_str()`, GTK+ always uses UTF-8.

File: src/gtk/mimetype.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    GIcon *gicon(g_content_type_get_icon(type));
+    if (gicon) {
```
Comment (vadz) [inline src/gtk/mimetype.cpp]: Please normalize indentation, i.e. put `{` on its own line.

File: src/gtk/mimetype.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+#include <gtk/gtk.h>
+
+#include "wx/gtk/private/string.h"
+
+wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString& mime)
+{
+    wxString icon;
+    wxGtkString type(g_content_type_from_mime_type(mime.mb_str()));
+
+    GIcon *gicon(g_content_type_get_icon(type));
```
Comment (vadz) [inline src/gtk/mimetype.cpp]: Using `wxGtkObject` from `include/wx/gtk/private/object.h` would be better than explicit `g_object_unref()` (below as well).

File: src/gtk/mimetype.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    return icon;
+}
+
+wxMimeTypesManagerImpl *wxGTKMimeTypesManagerFactory::CreateMimeTypesManagerImpl()
+{
+    return new wxGTKMimeTypesManagerImpl();
+}
+
+#endif
+  // wxUSE_MIMETYPE && wxUSE_FILE
```
Comment (vadz) [inline src/gtk/mimetype.cpp]: Please put the endif comment on the same line as endif itself.

**Comment by vadz** [comment]
I've left a lot of comments, but they're all rather minor and I don't see any problems with integrating this.

Out of curiosity, for which file types does this allow us to get the icons? I.e. how reliable is this?

**Comment by Hanmac** [comment]
thanks i will apply the things later.

wxGtkObject doesn't work because the objects might be null (if icon is not found), and then it does crash when it does try to unref. fixing wxGtkObject to support null with also adding IsOk() check would be nice.

for what icons can it get:
that depends on the theme, but if you would do "image/xyz" it would fall back to generic image icon f possible. I will extend the sample that you can input a mimetype too.

and how reliable: 
if the icon exist for the theme it should find it. (because of limitations from wxIcon i do exclude SVG images)

**Comment by Hanmac** [comment]
vadz thanks for merging but it was not 100% finished yet,
the new files need to be added to the bakefiles.

**Comment by nerijus** [comment]
With yesterday's wx trunk (I was using a month earlier version before)
M segfaults when viewing message with embedded picture:

(gdb) bt
#0  0x00007ffff31ee09c in g_type_check_instance_is_fundamentally_a ()
    at /lib64/libgobject-2.0.so.0
#1  0x00007ffff31ce447 in g_object_unref () at /lib64/libgobject-2.0.so.0
#2  0x00007ffff6ed821d in wxGtkObject<_GtkIconInfo>::~wxGtkObject() (this=0x7fffffff9c00, __in_chrg=<optimized out>) at ../wxWidgets/include/wx/gtk/private/object.h:22
#3  0x00007ffff6ed7faa in wxGTKMimeTypesManagerImpl::GetIconFromMimeType(wxString const&) (this=0x120f8c0, mime=...) at ../wxWidgets/src/gtk/mimetype.cpp:40
#4  0x00007ffff64ffbfa in wxMimeTypesManagerImpl::LoadXDGGlobs(wxString const&) (this=0x120f8c0, filename=...) at ../wxWidgets/src/unix/mimetype.cpp:265
#5  0x00007ffff6500e71 in wxMimeTypesManagerImpl::Initialize(int, wxString const&) (this=0x120f8c0, mailcapStyles=15, sExtraDir=...) at ../wxWidgets/src/unix/mimetype.cpp:534
#6  0x00007ffff6500dd4 in wxMimeTypesManagerImpl::InitIfNeeded() (this=0x120f8c0)
    at ../wxWidgets/src/unix/mimetype.cpp:518
#7  0x00007ffff6502c8b in wxMimeTypesManagerImpl::GetFileTypeFromMimeType(wxString const&) (this=0x120f8c0, mimeType=...) at ../wxWidgets/src/unix/mimetype.cpp:885
#8  0x00007ffff647883c in wxMimeTypesManager::GetFileTypeFromMimeType(wxString const&) (this=0x120fcc0, mimeType=...) at ../wxWidgets/src/common/mimecmn.cpp:662
#9  0x00000000004eec1c in wxIconManager::GetIconFromMimeType(wxString const&, wxString const&) (this=0xcf2950, type=..., ext=...) at .src/gui/wxIconManager.cpp:604
#10 0x00000000004885f7 in MessageView::ShowAttachment(MimePart const*) (this=0xeeb400, mimepart=0x1208880) at .src/classes/MessageView.cpp:2117
#11 0x0000000000488c54 in MessageView::ShowImage(MimePart const*) (this=0xeeb400, mimepart=0x1208880) at .src/classes/MessageView.cpp:2227
#12 0x00000000004892f8 in MessageView::ShowPart(MimePart const*) (this=0xeeb400, mimepart=0x1208880) at .src/classes/MessageView.cpp:2321
#13 0x000000000048b365 in MessageView::ProcessPart(MimePart const*, MessageView::MimePartAction) (this=0xeeb400, mimepart=0x1208880, action=MessageView::Part_Show)
    at .src/classes/MessageView.cpp:2774
#14 0x00000000004897db in MessageView::ProcessAllNestedParts(MimePart const*, MessageView::MimePartAction) (this=0xeeb400, mimepart=0x1208e00, action=MessageView::Part_Show)
    at .src/classes/MessageView.cpp:2381
#15 0x000000000048b191 in MessageView::ProcessMultiPart(MimePart const*, wxString const&, MessageView::MimePartAction) (this=0xeeb400, mimepart=0x1208e00, subtype=..., action=MessageView::Part_Show) at .src/classes/MessageView.cpp:2726
#16 0x000000000048b29f in MessageView::ProcessPart(MimePart const*, MessageView::MimePartAction) (this=0xeeb400, mimepart=0x1208e00, action=MessageView::Part_Show)
    at .src/classes/MessageView.cpp:2740
#17 0x000000000048c32f in MessageView::DisplayMessageInViewer() (this=0xeeb400)
    at .src/classes/MessageView.cpp:3137
#18 0x000000000048c19e in MessageView::Update() (this=0xeeb400)
    at .src/classes/MessageView.cpp:3081
#19 0x000000000049103d in MessageView::DoShowMessage(Message*) (this=0xeeb400, mailMessage=
    0x1207bf0) at .src/classes/MessageView.cpp:3978
#20 0x0000000000510eeb in wxMessageView::DoShowMessage(Message*) (this=0xeeb400, mailMessage=0x1207bf0) at .src/gui/wxMessageView.cpp:225
#21 0x00000000004850d2 in MessageView::OnASFolderResultEvent(MEventASFolderResultData&) (this=0xeeb400, event=...) at .src/classes/MessageView.cpp:1132
#22 0x0000000000484ee4 in MessageView::OnMEvent(MEventData&) (this=0xeeb400, event=...)
    at .src/classes/MessageView.cpp:1079
#23 0x0000000000496d07 in MEventManager::Dispatch(MEventData*) (dataptr=0x10a0360)
    at .src/classes/MEvent.cpp:196
#24 0x0000000000496954 in MEventManager::ForceDispatchPending() () at .src/classes/MEvent.cpp:146
#25 0x0000000000532f93 in wxFolderView::ShowFolder(MailFolder*) (this=0xef5a20, mf=0x1075bd0)
    at .src/gui/wxFolderView.cpp:4031
#26 0x0000000000533c1b in wxFolderView::OpenFolder(MFolder*, bool) (this=0xef5a20, folder=
    0xc2dfe0, readonly=false) at .src/gui/wxFolderView.cpp:4200
#27 0x00000000004e54ad in wxMainFrame::OpenFolder(MFolder*, bool) (this=
    0xccbdb0, pFolder=0xc2dfe0, readonly=false) at .src/gui/wxMainFrame.cpp:784
#28 0x00000000004a0fcf in MAppBase::ContinueStartup() (this=0xc48ae0)
    at .src/classes/MApplication.cpp:280
#29 0x00000000004a2a47 in MAppBase::OnStartup() (this=0xc48ae0)
    at .src/classes/MApplication.cpp:628
#30 0x00000000004b7507 in wxMApp::OnInit() (this=0xc48830) at .src/gui/wxMApp.cpp:1161
#31 0x00000000004be973 in wxAppConsoleBase::CallOnInit() (this=0xc48830)
    at /a/M/wxWindows.31/wxWidgets/include/wx/app.h:93
#32 0x00007ffff6465bcb in wxEntry(int&, wchar_t**) (argc=@0x7ffff67d5130: 1, argv=0xc45080)
    at ../wxWidgets/src/common/init.cpp:487

Regards,
Nerijus

**Comment by nerijus** [comment]
Segfault is gone if I revert this.

**Comment by Hanmac** [comment]
@nerijus : what is your gtk version and other stuff? because i couldnt reproduce it with gtk3.

@vadz: i am currently on another tiny patch for it setting the Description under gtk.
but it does return a localized version like "*.png" returns "PNG-Bild" in German. (or the current user language)

**Comment by nerijus** [comment]
wx is built with gtk2, not gtk3.

**Comment by Hanmac** [comment]
@nerijus: whats why i did ask, because my wx is build with gtk3. so it might be a compatiblity error between this two gtk versions.

otherwise you might look for what mimetype it does crash for you so we might look if we can repoduce it too.

what are your other System Settings, what is your Operating System?

**Comment by nerijus** [comment]
gtk2 is still used more than gtk3 with wx IMHO. So you should probably test your changes with it too.

**Comment by nerijus** [comment]
Fedora 23, mimetype is  IMAGE/PNG.

**Comment by Hanmac** [comment]
yeah its a bug/incompatiblity error between gtk2/gtk3 
wxGtkObject can't be used because of different free function.

@vadz  i also added the the GetDescription into it or do you want it as extra patch?

patch:
```diff
diff --git a/include/wx/gtk/mimetype.h b/include/wx/gtk/mimetype.h
index 4e5f144..a60517b 100644
--- a/include/wx/gtk/mimetype.h
+++ b/include/wx/gtk/mimetype.h
@@ -18,6 +18,7 @@ class WXDLLIMPEXP_CORE wxGTKMimeTypesManagerImpl : public wxMimeTypesManagerImpl
 {
 protected:
     wxString GetIconFromMimeType(const wxString& mime) wxOVERRIDE;
+    wxString GetDescriptionFromMimeType(const wxString& mime) wxOVERRIDE;
 };
 
 
diff --git a/include/wx/unix/mimetype.h b/include/wx/unix/mimetype.h
index 1436dd1..e2cda2d 100644
--- a/include/wx/unix/mimetype.h
+++ b/include/wx/unix/mimetype.h
@@ -100,6 +100,7 @@ protected:
                        const wxString& strDesc);
 
     virtual wxString GetIconFromMimeType(const wxString& mime);
+    virtual wxString GetDescriptionFromMimeType(const wxString& mime);
 
     // give it access to m_aXXX variables
     friend class WXDLLIMPEXP_FWD_BASE wxFileTypeImpl;
diff --git a/src/gtk/mimetype.cpp b/src/gtk/mimetype.cpp
index 31ee984..516548c 100644
--- a/src/gtk/mimetype.cpp
+++ b/src/gtk/mimetype.cpp
@@ -19,6 +19,12 @@
 #include "wx/gtk/private/string.h"
 #include "wx/gtk/private/object.h"
 
+wxString wxGTKMimeTypesManagerImpl::GetDescriptionFromMimeType(const wxString& mime)
+{
+    wxGtkString type(g_content_type_from_mime_type(mime.utf8_str()));
+    return wxString::FromUTF8(g_content_type_get_description(type));
+}
+
 wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString& mime)
 {
     wxGtkString type(g_content_type_from_mime_type(mime.utf8_str()));
@@ -31,17 +37,24 @@ wxString wxGTKMimeTypesManagerImpl::GetIconFromMimeType(const wxString& mime)
     if ( !theme )
         return wxString();
 
-    wxGtkObject<GtkIconInfo> giconinfo(gtk_icon_theme_lookup_by_gicon
-                                       (
+    GtkIconInfo *giconinfo = gtk_icon_theme_lookup_by_gicon(
                                             theme,
                                             gicon,
                                             256,
                                             GTK_ICON_LOOKUP_NO_SVG
-                                       ));
+                                       );
     if ( !giconinfo )
         return wxString();
 
-    return wxString::FromUTF8(gtk_icon_info_get_filename(giconinfo));
+    wxString icon = wxString::FromUTF8(gtk_icon_info_get_filename(giconinfo));
+
+#ifdef __WXGTK3__
+    g_object_unref(giconinfo);
+#else
+    gtk_icon_info_free(giconinfo);
+#endif
+
+    return icon;
 }
 
 wxMimeTypesManagerImpl *wxGTKMimeTypesManagerFactory::CreateMimeTypesManagerImpl()
diff --git a/src/unix/mimetype.cpp b/src/unix/mimetype.cpp
index 9e46216..52dc07e 100644
--- a/src/unix/mimetype.cpp
+++ b/src/unix/mimetype.cpp
@@ -263,8 +263,9 @@ void wxMimeTypesManagerImpl::LoadXDGGlobs(const wxString& filename)
        exts.Add( ext );
 
        wxString icon = GetIconFromMimeType(mime);
+       wxString description = GetDescriptionFromMimeType(mime);
 
-       AddToMimeData(mime, icon, NULL, exts, wxEmptyString, true );
+       AddToMimeData(mime, icon, NULL, exts, description, true );
     }
 }
 
@@ -694,6 +695,12 @@ wxString wxMimeTypesManagerImpl::GetIconFromMimeType(const wxString& WXUNUSED(mi
     return wxString();
 }
 
+wxString wxMimeTypesManagerImpl::GetDescriptionFromMimeType(const wxString& WXUNUSED(mime))
+{
+    return wxString();
+}
+
+
 bool wxMimeTypesManagerImpl::DoAssociation(const wxString& strType,
                                            const wxString& strIcon,
                                            wxMimeTypeCommands *entry,
```

**Comment by vadz** [comment]
Damn, looks like I really jumped the gun with applying this, sorry :-(

Looking at the docs, it's wrong to use `g_object_unref()` with `GtkIconInfo`, we need to use `gtk_icon_info_free()` instead.

**Comment by vadz** [comment]
@Hanmac our comments crossed, sorry. I'll apply the bug fix part, please open another PR for the description. TIA!

**Comment by vadz** [comment]
@nerijus Thanks for testing/reporting this! It should be fixed by the commit above (3c4f51df70336730aa992d26548f75e2d09d7e6d), please let me know if you still have any problems.

**Comment by nerijus** [comment]
It is fixed, thanks.

**Pull request closed without merging**